packages/react-router/src/link.tsx

```diff
@@ -84,67 +84,78 @@ export type RemoveLeadingSlashes<T> = T extends `/${infer R}`
   ? RemoveLeadingSlashes<R>
   : T
 
+export type ResolvePaths<TRouteTree extends AnyRoute, TSearchPath> =
+  RouteByPath<TRouteTree, RemoveTrailingSlashes<TSearchPath>> extends never
+    ? RoutePaths<TRouteTree>
+    : RoutePaths<RouteByPath<TRouteTree, RemoveTrailingSlashes<TSearchPath>>>
+
 export type SearchPaths<
-  TPaths,
+  TRouteTree extends AnyRoute,
   TSearchPath extends string,
+  TPaths = ResolvePaths<TRouteTree, TSearchPath>,
 > = TPaths extends `${RemoveTrailingSlashes<TSearchPath>}/${infer TRest}`
   ? TRest
   : never
 
 export type SearchRelativePathAutoComplete<
+  TRouteTree extends AnyRoute,
   TTo extends string,
   TSearchPath extends string,
-  TPaths,
-  TSearchedPaths = SearchPaths<TPaths, TSearchPath>,
-> = TSearchedPaths extends string ? `${TTo}/${TSearchedPaths}` : never
+> = `${TTo}/${SearchPaths<TRouteTree, TSearchPath>}`
```

Comment:
This was a silly mistake by me before. Template literal types automatically distribute. Its a small optimization but it does help a bit